### PR TITLE
add import tool

### PR DESCRIPTION
**Pull request by peizhou298**
## Description
Describe the changes in this PR.

## Related issues
Addresses [issue #].

## Testing
Describe how this change was tested.

## FHIR Team Checklist
- [ ] **Update the title** of the PR to be succinct and less than 50 characters
- [ ] **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- [ ] Tag the PR with the type of update: **Bug**, **Dependencies**, **Enhancement**, or **New-Feature**
- [ ] Tag the PR with **Azure API for FHIR** if this will release to the managed service
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

File: ImportTool/ImportTool.csproj [hunk 1]
```diff
@@ -0,0 +1,15 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Hl7.Fhir.R4" Version="3.4.0" />
```
Comment (tongwu-msft) [inline ImportTool/ImportTool.csproj:9]: We need to support stu3/r4/r5.

File: ImportTool/Program.cs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+using System;
+using System.Threading.Tasks;
+using Microsoft.Azure.Storage;
+using Mono.Options;
+
+namespace ImportTool
+{
+    public static class Program
+    {
+        private const string _localConnectionString = "DefaultEndpointsProtocol=http;AccountName=devstoreaccount1;AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==;BlobEndpoint=http://127.0.0.1:10000/devstoreaccount1;";
```
Comment (tongwu-msft) [inline ImportTool/Program.cs:15]: remove credential

File: ImportTool/Program.cs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        private static Config _config = new Config
+        {
+            StorageConnectionString = _localConnectionString,
+            SplitSizeInBytes = _defaultSplitSizeInBytes,
+            MaxFileNumber = _defaultMaxFileNumber,
+        };
+
+        public static async Task Main(string[] args)
+        {
+            string storageConnectionString = Environment.GetEnvironmentVariable("AZURE_STORAGE_CONNECTION_STRING");
```
Comment (tongwu-msft) [inline ImportTool/Program.cs:28]: Use cmd inputs for account & key.

File: ImportTool/Program.cs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            _config.StorageConnectionString = storageConnectionString ?? _config.StorageConnectionString;
+
+            bool show_help = false;
```
Comment (tongwu-msft) [inline ImportTool/Program.cs:31]: use c# format.

File: ImportTool/Program.cs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            bool generate_request = false;
+            bool split_file = false;
+            string prefix = string.Empty;
+
+            var option = new OptionSet()
+            {
+                {
+                    "c|connectionString=", "the {CONNECTIONSTRING} of source blob.",
+#pragma warning disable SA1501 // Statement should not be on a single line
```
Comment (tongwu-msft) [inline ImportTool/Program.cs:40]: we'd better follow CA warning

File: ImportTool/RequestGenerator.cs [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+                        inputPart.Add(Tuple.Create("type", (Base)new FhirString(resourceType)));
+                        inputPart.Add(Tuple.Create("url", (Base)new FhirUri(segment.Uri)));
+                        string etag = segment.Properties.ETag.Replace("\"", string.Empty, StringComparison.OrdinalIgnoreCase);
+                        inputPart.Add(Tuple.Create("etag", (Base)new FhirString(etag)));
+                        parameters.Add("input", inputPart);
+                        succeed_count++;
+                    }
+                    catch (Exception oex)
+                    {
+                        Console.WriteLine($"Error reslove file {segment.Uri} due to exception {oex.Message}");
```
Comment (tongwu-msft) [inline ImportTool/RequestGenerator.cs:76]: use logger for log message

Comment (tongwu-msft) [reply ImportTool/RequestGenerator.cs:76]: we should throw exception to customer

File: ImportTool/RequestGenerator.cs [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        private static async Task<string> GetResourceType(CloudBlockBlob cloudBlockBlob)
+        {
+            using (var stream = await cloudBlockBlob.OpenReadAsync())
+            {
+                char[] buffer = ArrayPool<char>.Shared.Rent(128);
+                try
+                {
+                    using (StreamReader reader = new StreamReader(stream))
+                    {
+                        reader.Read(buffer, 0, 128);
```
Comment (tongwu-msft) [inline ImportTool/RequestGenerator.cs:97]: we can read first line

File: ImportTool/RequestGenerator.cs [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+                    }
+                    catch (Exception oex)
+                    {
+                        Console.WriteLine($"Error reslove file {segment.Uri} due to exception {oex.Message}");
+                    }
+
+                    count++;
+                }
+            }
+            while (continuationToken != null && count < maxFileNumber);
```
Comment (tongwu-msft) [inline ImportTool/RequestGenerator.cs:82]: we can check for each file, instead check it after all files generated.

File: ImportTool/RequestGenerator.cs [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+            int succeed_count = 0;
+
+            BlobContinuationToken continuationToken = null;
+            do
+            {
+                var segments = await cloudBlobClient.ListBlobsSegmentedAsync(
+                    prefix: prefix,
+                    useFlatBlobListing: true,
+                    blobListingDetails: BlobListingDetails.Metadata,
+                    maxResults: Math.Min(maxFileNumber, 5000),
```
Comment (tongwu-msft) [inline ImportTool/RequestGenerator.cs:56]: we need to check .ndjson

File: ImportTool/RequestGenerator.cs [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+                Console.WriteLine($"Succeed files: {succeedCount} put as input.");
+            }
+
+            return new FhirJsonSerializer().SerializeToString(parameters);
+        }
+
+        private static async Task<Tuple<int, int>> AddInputPartsFromBlobs(CloudBlobClient cloudBlobClient, string prefix, int maxFileNumber, Parameters parameters)
+        {
+            int count = 0;
+            int succeed_count = 0;
```
Comment (tongwu-msft) [inline ImportTool/RequestGenerator.cs:47]: we do not need to handle failure, just fail fast.

File: ImportTool/RequestGenerator.cs [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+using Hl7.Fhir.Model;
+using Hl7.Fhir.Serialization;
+using Microsoft.Azure.Storage;
+using Microsoft.Azure.Storage.Blob;
+
+namespace ImportTool
+{
+    public static class RequestGenerator
+    {
+        public static async Task<string> GenerateImportRequest(string connectionString, string prefix, int maxFileNumber)
```
Comment (tongwu-msft) [inline ImportTool/RequestGenerator.cs:22]: why we have this limit?

File: ImportTool/Program.cs [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+using Microsoft.Azure.Storage;
+using Mono.Options;
+
+namespace ImportTool
+{
+    public static class Program
+    {
+        private const string _localConnectionString = "DefaultEndpointsProtocol=http;AccountName=devstoreaccount1;AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==;BlobEndpoint=http://127.0.0.1:10000/devstoreaccount1;";
+        private const long _defaultSplitSizeInBytes = 200 * 1024 * 1024L;
+        private const int _defaultMaxFileNumber = 2000;
```
Comment (tongwu-msft) [inline ImportTool/Program.cs:17]: constants should not at main class.

File: ImportTool/Program.cs [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+                option.Parse(args);
+
+                if (generate_request)
+                {
+                    await RequestGenerator.GenerateImportRequest(_config.StorageConnectionString, prefix, _config.MaxFileNumber);
+                }
+            }
+            catch (OptionException oe)
+            {
+                Console.WriteLine($"Error reslove arguments due to {oe.Message}");
```
Comment (tongwu-msft) [inline ImportTool/Program.cs:87]: we should throw all exception to customer

File: ImportTool/Program.cs (hunk 2 above)
Comment (peizhou298) [reply ImportTool/Program.cs:15]: It is a emunlator common connection string , but still let me replace it with a shorter one.

File: ImportTool/Program.cs (hunk 3 above)
Comment (peizhou298) [reply ImportTool/Program.cs:28]: Removed this line.

File: ImportTool/RequestGenerator.cs (hunk 7 above)
Comment (peizhou298) [reply ImportTool/RequestGenerator.cs:97]: First line may be very large, like imageStudy.

File: ImportTool/RequestGenerator.cs (hunk 9 above)
Comment (peizhou298) [reply ImportTool/RequestGenerator.cs:56]: Add regex match for that

File: ImportTool/RequestGenerator.cs (hunk 11 above)
Comment (peizhou298) [reply ImportTool/RequestGenerator.cs:22]: It is not a limit, user can choose how much files they want to set as input

File: ImportTool/Program.cs (hunk 12 above)
Comment (peizhou298) [reply ImportTool/Program.cs:17]: Will Remove it

File: ImportTool/Program.cs (hunk 13 above)
Comment (peizhou298) [reply ImportTool/Program.cs:87]: Will throw it.

File: ImportTool/Program.cs (hunk 5 above)
Comment (peizhou298) [reply ImportTool/Program.cs:40]: Fixed

File: ImportTool/RequestGenerator.cs (hunk 6 above)
Comment (peizhou298) [reply ImportTool/RequestGenerator.cs:76]: Fixed

File: ImportTool/Program.cs (hunk 4 above)
Comment (peizhou298) [reply ImportTool/Program.cs:31]: Fixed

File: ImportTool/ImportTool.csproj (hunk 1 above)
Comment (peizhou298) [reply ImportTool/ImportTool.csproj]: added shared project for fix

**Pull request closed without merging**